SAP Application Router Guide (#574)
SAP/cloud-sdk: docs/js/guides/how-to-use-the-approuter.mdx

---
title: Using the SAP Application Router with the SAP Cloud SDK
sidebar_label: Using the SAP Application Router
description: This article describes how you can use the SAP Application Router with the SAP Cloud SDK
keywords:
  - sap
  - cloud
  - sdk
  - approuter
  - multi-tenancy
  - cloud-foundry
  - JavaScript
  - TypeScript
---

# Using the SAP Application Router with the SAP Cloud SDK

> In this guide, we will show you how to use the SAP Application Router together with the SAP Cloud SDK.
> We will show you how to secure your application and utilize multi-tenancy for principal propagation with a nestJS example.

Reviewer: ```suggestion
In this guide, we will show you how to use the SAP Application Router together with the SAP Cloud SDK.
You will learn how to secure your application and configure multi-tenancy for principal propagation with an SDK-based application example powered by [nestJS](https://nestjs.com/).
```

> ## SAP Application Router
> When we want to enable multi-tenancy for our application, we use the SAP Application Router.
> The application router’s primary purpose is to be the single entry point of a microservice-based application and act as the application’s reverse proxy.
> Its responsibilities consist of dispatching requests to backend microservices, authenticating users, and serving static content.
> The application router checks if a given request has a valid JSON Web Token (JWT) when accessing a target service.
> If the request contains a valid JWT, the application router forwards the request to the target service; if the request does not contain a valid JWT, the user must authenticate himself.
> Since we use an Identity Provider (IdP) to authenticate ourselves, the request is redirected to an IdP and then redirected back to the application router. 

Reviewer: ```suggestion
## SAP Application Router

When we want to enable multi-tenancy for our application, we use the [SAP Application Router](https://www.npmjs.com/package/@sap/approuter).
The application router’s primary purpose is to be the single entry point of a microservice-based application and act as the application’s reverse proxy.
Its responsibilities consist of dispatching requests to backend microservices, authenticating users, and serving static content.
The application router checks if a given request has a valid JSON Web Token (JWT) when accessing a target service.
If the request contains a valid JWT, the application router forwards the request to the target service; if the request does not contain a valid JWT, the user must authenticate.
Since we use an Identity Provider (IdP) to authenticate, the request is redirected to an IdP where a user gets authenticated and then redirected back to the application router for passing further according to its desired destination.
```

Reviewer: A simple diagram here would have been great! Should we try to use mermaid?
I'll send you a code snippet.

> ### Approuter Setup
> To deploy our approuter in SAP BTP Cloud Foundry, we need to configure it first.

Reviewer: ```suggestion
### Approuter Setup

To deploy our approuter in SAP BTP Cloud Foundry, we need to configure it first.
```

Let's walk through the four files we need to use.

The `xs-security.json` file defines the security and deployment options for an application.
With the below example, we enable the `shared` tenant-mode for our xsuaa instance, which we need for multi-tenancy.
```json
{
  "xsappname": "approuter-scaffold",
  "tenant-mode": "shared"
}
```

In the `xs-app.json` we specify to which backend service a request is forwarded to, and how this request has to be authenticated.

> We can optionally also specify a specific `identityProvider` which is used for the authentication.

Reviewer: ```suggestion
We can optionally also specify a specific `identityProvider` that is used for the authentication.
```

{
  "name": "approuter",
  "dependencies": {
    "@sap/approuter": "*"
  },
  "scripts": {
    "start": "node node_modules/@sap/approuter/approuter.js"
  }
}
```

The manifest contains our approuter, as well as environment variables which our approuter needs for multi-tenancy.
As you see under `env`, we specify the `TENANT_HOST_PATTERN` and `destinations`.
The `destinations` are the destinations we use in our `xs-app.json` where we forward requests to. 
The `TENANT_HOST_PATTERN` is a regular expression that describes how a tenant name should be retrieved from the host.
We also have to bind the xsuaa which we configured with our `xs-security.json` to the approuter.

```yaml
applications:
  - name: approuter-scaffold-approuter
    routes:
      - route: >-
          approuter-scaffold-apps.cfapps.sap.hana.ondemand.com
    path: .
    memory: 128M
    buildpacks:
      - nodejs_buildpack
    env:
      TENANT_HOST_PATTERN: >-
        "approuter-scaffold-(.*).cfapps.sap.hana.ondemand.com"
      destinations: >-
        [{"name":"approuter-scaffold","url":"approuter-scaffold.cfapps.sap.hana.ondemand.com","forwardAuthToken":true}]
    services:
      - approuter-scaffold-xsuaa
```

> ## Securing your Application
> To secure our application endpoints, we utilize the [passport](http://www.passportjs.org/packages/passport-jwt) library.

Reviewer: ```suggestion
## Securing your Application

To secure our application endpoints, we utilize the [passport](http://www.passportjs.org/packages/passport-jwt) library.
```
